src/beamlime/applications/_nexus_helpers.py

```diff
@@ -1,155 +1,378 @@
 # SPDX-License-Identifier: BSD-3-Clause
 # Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
-from typing import Iterable, Mapping, Optional, Tuple
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    Literal,
+    Mapping,
+    Optional,
+    Sequence,
+    TypedDict,
+)
 
 import numpy as np
 
 
-def _dtype(value):
-    if hasattr(value, 'dtype'):
-        return str(value.dtype)
-    if isinstance(value, str):
-        return 'string'
-    if isinstance(value, int):
-        return 'int'
-    if isinstance(value, float):
-        return 'float'
-    return 'unknown'
-
-
-def _init_dataset(group, field, value):
-    dataset = find_nexus_structure(group, (field,))
-    if dataset is None:
-        dataset = dict(
-            module='dataset',
-            config=dict(
-                name=field,
-                dtype=_dtype(value),
-                values=value,
-            ),
+class NexusDatasetDict(TypedDict):
+    """``dataset`` module structure in the nexus json format."""
+
+    module: str
+    config: dict[str, Any]
+    attributes: list[Mapping[str, Any]]
+
+
+NexusPath = tuple[str | None, ...]
+
+
+class NexusGroupDict(TypedDict):
+    """A nexus group that holds the data module place holder or datasets.
+
+    It was named as a group, not a module
+    since the module place holder is one of the children of it.
+    """
+
+    children: list[Mapping]
+    """A module place holder or datasets."""
+    name: str
+    """Name of the group."""
+
+
+NexusGroupDictStore = dict[NexusPath, NexusGroupDict]
+"""A dictionary to store nexus groups for merging data pieces."""
+NexusDictStructure = dict
+"""A whole/partial nexus structure used for the data acquisition."""
+
+
+def create_dataset(
+    *, name: str, dtype: str, initial_values: Any, unit: str | None = None
+) -> NexusDatasetDict:
+    """Creates a dataset according to the arguments."""
+    dataset: NexusDatasetDict = {
+        "module": "dataset",
+        "config": {
+            "name": name,
+            "dtype": dtype,
+            "values": initial_values,
+        },
+        "attributes": [],
+    }
+    if unit is not None:
+        attrs = dataset["attributes"]
+        attrs.append(
+            {
+                "name": "units",
+                "dtype": "string",
+                "values": unit,
+            }
         )
-        group['children'].append(dataset)
     return dataset
 
 
-def _merge_ev44(group, message):
-    for field, value in message.items():
-        if value is None:  # Could be monitor without pixel_id
-            continue
-        dataset = _init_dataset(group, field, value)
-        if field in ('time_of_flight', 'pixel_id'):
-            dataset['config']['values'] = np.concatenate(
-                (dataset['config']['values'], value)
+ModuleNameType = Literal["ev44"]  # "f144", "tdct" will be added in the future
+"""Name of the module that is supported by beamlime."""
+
+DeserializedMessage = Mapping
```

Comment:
I don't think we need to specify that it is a deserialized message. It's enough to say that it is a message, or maybe `StreamMessage` if we want to specify that it is part of a module strem.

Reply:
I find it confusing because there are three types of messages:
- deserialized `flat_buffer`
- kafka message
- message for a beamlime message router

Reply:
I thought it's worth mention that it's deserialized flat buffer.

Reply:
Okay then let's keep it as it is